src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -326,69 +323,100 @@ where
     );
 
     let name = format!("SourceGenericStats({})", source_id);
-    let mut builder = AsyncOperatorBuilder::new(name, scope.clone());
-
-    let (data_output, data) = builder.new_output::<CapacityContainerBuilder<_>>();
-    let (progress_output, derived_progress) = builder.new_output::<CapacityContainerBuilder<_>>();
-    let mut data_input = builder.new_input_for_many(
-        &input_data.inner,
-        Pipeline,
-        [&data_output, &progress_output],
-    );
-    let (health_output, derived_health) = builder.new_output::<CapacityContainerBuilder<_>>();
+    let mut builder = OperatorBuilderRc::new(name, scope.clone());
+
+    let (_, derived_progress) = builder.new_output::<CapacityContainerBuilder<_>>();
+    let (mut health_output, derived_health) = builder.new_output::<CapacityContainerBuilder<_>>();
+
+    let mut export_collections = BTreeMap::new();
+    let mut export_handles = vec![];
+    for (id, export) in exports {
+        let (export_output, new_export) = builder.new_output::<CapacityContainerBuilder<_>>();
+        // This input's frontier flow into the progress collection and the corresponding output
+        let mut connection = vec![Antichain::from_elem(Default::default()), Antichain::new()];
+        // No frontier implications for other outputs
+        for _ in 0..export_handles.len() {
+            connection.push(Antichain::new());
+        }
+        // Standard frontier implication for the corresponding output of this input
+        connection.push(Antichain::from_elem(Default::default()));
+        let export_input = builder.new_input_connection(&export.inner, Pipeline, connection);
+        export_handles.push((id, export_input, export_output));
+        let new_export: StackedCollection<G, Result<SourceMessage, DataflowError>> =
+            new_export.as_collection();
+        export_collections.insert(id, new_export);
```

Comment:
This loop is one that needs quite a bit of attention from the reviewer. This is an operator that simply passes through the data produced by the source and records some statistics and health related information as it sees the data passing by. Now that each output of a source is a separate stream this is an operator with N inputs and N outputs. The operator connections must be such that the i-th input is only connected to the i-th output, plus the progress collection.

It is important that we get this right since the failure mode will be hard to spot.